python/mspasspy/global_history/manager.py

```diff
@@ -1,21 +1,179 @@
 import os
 import yaml
 import pymongo
-import pyspark
 import collections
-import dask.bag as daskbag
 import json
 
+try:
+    import dask.bag as daskbag
+    import pyspark
+except:
+    pass
 from bson.objectid import ObjectId
 from mspasspy.ccore.utility import MsPASSError, AntelopePf
 from mspasspy.util.converter import AntelopePf2dict
 from datetime import datetime
 from dill.source import getsource
+import functools
 
 import mspasspy.algorithms.signals as signals
 from mspasspy.global_history.ParameterGTree import ParameterGTree, parameter_to_GTree
 
 
+def mspass_normal_map(
+    data,
+    func,
+    *args,
+    global_history=None,
+    object_history=False,
+    alg_id=None,
+    alg_name=None,
+    parameters=None,
+    **kwargs
+):
+    """
+     This decorator method performs map function in Python. Instead of performing the normal map function,
+     if user provides global history manager, alg_id(optional), alg_name(optional) and parameters(optional)
+     as input, the global history manager will log down the usage of the algorithm. Also, if user set
+     object_history to be True, then each mspass object in this map function will save the object level history.
+
```

Comment:
definition of data is missing. Also, in the map function, it doesn't need `args` and `kwargs`.